### PR TITLE
When every import is unused, error on the entire import declaration

### PR DESCRIPTION
**Pull request by andy-ms**
Part of https://github.com/Microsoft/TypeScript/pull/22361#issuecomment-371230932

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -21682,18 +21682,45 @@ namespace ts {
...
                         }
                     }
                 });
+
+                while (unusedImports.length) {
+                    const decl = unusedImports.pop()!;
+                    const importClause = decl.kind === SyntaxKind.ImportClause ? decl : decl.kind === SyntaxKind.NamespaceImport ? decl.parent : decl.parent.parent;
+                    if (!forEachImportedDeclaration(importClause, d => d !== decl && !contains(unusedImports, d))) {
+                        forEachImportedDeclaration(importClause, d => unorderedRemoveItem(unusedImports, d));
+                        error(importClause.parent, Diagnostics.No_import_of_0_is_used, showModuleSpecifier(importClause.parent));
```
Comment (mhegazy) [inline src/compiler/checker.ts]: 1. if there is only one import e.g. default or one named import, the other error is superior. i would suggests such extending the span and using the error message with he name of the import.

2. I think the new error message is inaccurate. we just know that the import declaration is not needed, but there may be other import declarations for the same module that are still in use. so possibly remove the name of the module from the error message and say `Import declaration unused` or `all imports in import declaration are unused`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -21682,18 +21682,45 @@ namespace ts {
...
-                    if (!local.isReferenced && !local.exportSymbol) {
-                        for (const declaration of local.declarations) {
-                            if (!isAmbientModule(declaration)) {
-                                errorUnusedLocal(declaration, symbolName(local));
-                            }
+                    if (local.isReferenced || local.exportSymbol) return;
+                    for (const declaration of local.declarations) {
+                        if (isAmbientModule(declaration)) continue;
+                        if (isImportedDeclaration(declaration)) {
+                            unusedImports.push(declaration);
```
Comment (amcasey) [inline src/compiler/checker.ts]: Personally, I find the forEach code below a little hard to follow.  Another way to express this might be to build a multi-map from import clauses to the unused declarations they contain.  Then, for each import clause in the map, you could either report a single diagnostic or a list of diagnostics according to whether or not the import clause contained a declaration not in the corresponding unused list.  I don't feel strong about this.

File: src/services/codefixes/fixUnusedIdentifier.ts [hunk 3]
```diff
@@ -38,14 +47,24 @@ namespace ts.codefix {
                     }
                     break;
                 case fixIdDelete:
-                    tryDeleteDeclaration(changes, sourceFile, token);
+                    if (diag.code === Diagnostics.No_import_of_0_is_used.code) {
+                        changes.deleteNode(sourceFile, getImportDeclarationAtErrorLocation(token));
```
Comment (amcasey) [inline src/services/codefixes/fixUnusedIdentifier.ts]: Which adjacent trivia does this delete?

File: tests/cases/fourslash/unusedImportsFS_entireImportDeclaration.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+/// <reference path='fourslash.ts' />
+
+// @noUnusedLocals: true
+// @Filename: /a.ts
+////import a, { b } from "mod";
+
+verify.codeFix({
+    description: "Remove declaration for: 'mod'",
```
Comment (amcasey) [inline tests/cases/fourslash/unusedImportsFS_entireImportDeclaration.ts]: This wording is confusing.  What about something like "Remove import from 'mod'"?

File: src/services/codefixes/fixUnusedIdentifier.ts (hunk 3 above)
Comment (andy-ms) [reply src/services/codefixes/fixUnusedIdentifier.ts]: Looks like it removes leading but not trailing trivia by default.

File: src/compiler/checker.ts (hunk 2 above)
Comment (andy-ms) [reply src/compiler/checker.ts]: Thanks, helped me remove the `count` variable.

**Comment by amcasey** [comment]
@andy-ms looks like you might need to regenerate baselines - there's a conflict with your expanded range change.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -21682,18 +21682,50 @@ namespace ts {
...
+                            unusedImports.add(String(getNodeId(importClauseFromImported(declaration))), declaration);
+                        }
+                        else {
+                            errorUnusedLocal(declaration, symbolName(local));
                         }
                     }
                 });
+
+                unusedImports.forEach(unuseds => {
+                    const importClause = importClauseFromImported(first(unuseds)); // others will have the same clause
```
Comment (amcasey) [inline src/compiler/checker.ts]: Why recompute this?  Isn't it the map key?

Comment (andy-ms) [reply src/compiler/checker.ts]: Unfortunately we don't support non-string keys for maps for compatibility with ES5. See `shimMap`.

Comment (andy-ms) [reply src/compiler/checker.ts]: We could always make the map value be a pair though...

Comment (amcasey) [reply src/compiler/checker.ts]: I must have been thinking of another context.  This makes sense, given that limitation.
